test/parallel/test-fs-promises-watch.js

```diff
@@ -0,0 +1,155 @@
+        clearInterval(interval);
+        interval = undefined;
+      }
+      assert.strictEqual(['rename', 'change'].includes(eventType), true);
+      assert.strictEqual(filename, testCase.fileName);
+      break;
+    }
+
+    // Waiting on it again is a non-op
+    // eslint-disable-next-line no-unused-vars
+    for await (const p of watcher) {
+      assert.fail('should not run');
+    }
+  }
+
+  // Long content so it's actually flushed. toUpperCase so there's real change.
+  const content2 = Date.now() + testCase.fileName.toUpperCase().repeat(1e4);
+  interval = setInterval(() => {
+    fs.writeFileSync(testCase.filePath, '');
+    fs.writeFileSync(testCase.filePath, content2);
+  }, 100);
+
+  test().then(common.mustCall());
+}
+
+assert.rejects(
+  // eslint-disable-next-line no-unused-vars
+  async () => { for await (const _ of watch(1)) {} },
+  { code: 'ERR_INVALID_ARG_TYPE' });
+
+assert.rejects(
+  // eslint-disable-next-line no-unused-vars
+  async () => { for await (const _ of watch(__filename, 1)) {} },
+  { code: 'ERR_INVALID_ARG_TYPE' });
+
+assert.rejects(
+  // eslint-disable-next-line no-unused-vars
+  async () => { for await (const _ of watch('', { persistent: 1 })) {} },
+  { code: 'ERR_INVALID_ARG_TYPE' });
+
+assert.rejects(
+  // eslint-disable-next-line no-unused-vars
+  async () => { for await (const _ of watch('', { recursive: 1 })) {} },
+  { code: 'ERR_INVALID_ARG_TYPE' });
+
+assert.rejects(
+  // eslint-disable-next-line no-unused-vars
+  async () => { for await (const _ of watch('', { encoding: 1 })) {} },
+  { code: 'ERR_INVALID_ARG_VALUE' });
+
+assert.rejects(
+  // eslint-disable-next-line no-unused-vars
+  async () => { for await (const _ of watch('', { signal: 1 })) {} },
+  { code: 'ERR_INVALID_ARG_TYPE' });
+
+(async () => {
+  const ac = new AbortController();
+  const { signal } = ac;
+  setImmediate(() => ac.abort());
+  try {
+    // eslint-disable-next-line no-unused-vars
+    for await (const _ of watch(__filename, { signal })) {}
+  } catch (err) {
+    assert.strictEqual(err.name, 'AbortError');
+  }
+})().then(common.mustCall());
+
+tmpdir.refresh();
+
+for (const testCase of kCases) {
+  let interval;
+  if (testCase.shouldSkip) continue;
+  fs.mkdirSync(testCase.dirPath);
+  // Long content so it's actually flushed.
+  const content1 = Date.now() + testCase.fileName.toLowerCase().repeat(1e4);
+  fs.writeFileSync(testCase.filePath, content1);
+
+  const ac = new AbortController();
+
+  async function test() {
+    const { signal } = ac;
+    signal.addEventListener(
+      'abort',
+      () => clearInterval(interval),
+      { once: true });
+    const watcher = watch(testCase[testCase.field], { signal });
+    assert.rejects(watcher.next(), {
+      name: 'AbortError'
+    });
+  }
+
+  // Long content so it's actually flushed. toUpperCase so there's real change.
+  const content2 = Date.now() + testCase.fileName.toUpperCase().repeat(1e4);
+  interval = setInterval(() => {
```

Comment:
I am not sure why this is done in a setInterval that immediately gets removed when aborted rather than a setTimeout

Reply:
Not sure lol.. I was following the pattern in the existing test-fs-watch ... I have a todo to revisit it.